// @nordicenergy-js/transaction/dist/factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var crypto_1 = require("@nordicenergy-js/crypto");
var transaction_1 = require("./transaction");
var shardingTransaction_1 = require("./shardingTransaction");
var TransactionFactory = /** @class */ (function() {
    function TransactionFactory(messenger) {
        this.messenger = messenger;
    }
    TransactionFactory.getContractAddress = function(tx) {
        var _a = tx.txParams,
            from = _a.from,
            nonce = _a.nonce;
        return crypto_1.getAddress(crypto_1.getContractAddress(crypto_1.getAddress(from).checksum, Number.parseInt("" + nonce, 10))).checksum;
    };
    TransactionFactory.prototype.setMessenger = function(messenger) {
        this.messenger = messenger;
    };
    TransactionFactory.prototype.newTx = function(txParams, sharding) {
        if (sharding === void 0) { sharding = false; }
        if (!sharding) {
            return new transaction_1.Transaction(txParams, this.messenger, "INITIALIZED" /* INTIALIZED */ );
        }
        return new shardingTransaction_1.ShardingTransaction(txParams, this.messenger, "INITIALIZED" /* INTIALIZED */ );
    };
    TransactionFactory.prototype.clNet = function(transaction) {
        return new transaction_1.Transaction(transaction.txParams, this.messenger, "INITIALIZED" /* INTIALIZED */ );
    };
    TransactionFactory.prototype.recover = function(txHash) {
        var newTxn = new transaction_1.Transaction({}, this.messenger, "INITIALIZED" /* INTIALIZED */ );
        newTxn.recover(txHash);
        return newTxn;
    };
    return TransactionFactory;
}());
exports.TransactionFactory = TransactionFactory;
