// @nordicenergy-js/core/dist/blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var network_1 = require("@nordicenergy-js/network");
var utils_1 = require("@nordicenergy-js/utils");
var crypto_1 = require("@nordicenergy-js/crypto");
var Blockchain = /** @class */ (function() {
    function Blockchain(messenger) {
        this.messenger = messenger;
    }
    Blockchain.prototype.setMessenger = function(messenger) {
        this.messenger = messenger;
    };
    Blockchain.prototype.getRpcResult = function(result) {
        if (result instanceof network_1.ResponseMiddleware) {
            return result.getRaw;
        } else {
            return result;
        }
    };
    /**
     *
     */
    Blockchain.prototype.getBalance = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getBalance" /* GetBalance */ , [crypto_1.getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockNumber = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_blockNumber" /* BlockNumber */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getBlockByHash = function(_a) {
        var blockHash = _a.blockHash,
            _b = _a.returnObject,
            returnObject = _b === void 0 ? true : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getBlockByHash" /* GetBlockByHash */ , [blockHash, returnObject], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getBlockByNumber = function(_a) {
        var _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.returnObject,
            returnObject = _c === void 0 ? true : _c,
            _d = _a.shardID,
            shardID = _d === void 0 ? this.messenger.currentShard : _d;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_e) {
                switch (_e.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getBlockByNumber" /* GetBlockByNumber */ , [blockNumber, returnObject], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _e.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockTransactionCountByHash = function(_a) {
        var blockHash = _a.blockHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getBlockTransactionCountByHash" /* GetBlockTransactionCountByHash */ , [blockHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockTransactionCountByNumber = function(_a) {
        var blockNumber = _a.blockNumber,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getBlockTransactionCountByNumber" /* GetBlockTransactionCountByNumber */ , [blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getTransactionByBlockHashAndIndex = function(_a) {
        var blockHash = _a.blockHash,
            index = _a.index,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getTransactionByBlockHashAndIndex" /* GetTransactionByBlockHashAndIndex */ , [blockHash, index], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionByBlockNumberAndIndex = function(_a) {
        var _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            index = _a.index,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getTransactionByBlockNumberAndIndex" /* GetTransactionByBlockNumberAndIndex */ , [blockNumber, index], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionByHash = function(_a) {
        var txnHash = _a.txnHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getTransactionByHash" /* GetTransactionByHash */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getTransactionReceipt = function(_a) {
        var txnHash = _a.txnHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getTransactionReceipt" /* GetTransactionReceipt */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getCxReceiptByHash = function(_a) {
        var txnHash = _a.txnHash,
            shardID = _a.shardID;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getCXReceiptByHash" /* GetCXReceiptByHash */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getCode = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getCode" /* GetCode */ , [crypto_1.getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.net_peerCount = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("net_peerCount" /* PeerCount */ , [], 'net', shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.net_version = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("net_version" /* NetVersion */ , [], 'net', shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getProtocolVersion = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_protocolVersion" /* ProtocolVersion */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getStorageAt = function(_a) {
        var address = _a.address,
            position = _a.position,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getStorageAt" /* GetStorageAt */ , [crypto_1.getAddress(address).checksum, position, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionCount = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getTransactionCount" /* GetTransactionCount */ , [crypto_1.getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getShardingStructure = function() {
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_getShardingStructure" /* GetShardingStructure */ , [], this.messenger.chainPrefix)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.sendTransaction = function(transaction) {
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        if (!transaction.isSigned() || !transaction) {
                            throw new Error('transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , this.messenger.send("ngy_sendRawTransaction" /* SendRawTransaction */ , [transaction.getRawTransaction()], this.messenger.chainPrefix, typeof transaction.txParams.shardID === 'string' ?
                            Number.parseInt(transaction.txParams.shardID, 10) :
                            transaction.txParams.shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.sendRawTransaction = function(transaction) {
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var _a, txn, result;
            return tslib_1.__generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        if (!transaction.isSigned() || !transaction) {
                            throw new Error('transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , transaction.sendTransaction()];
                    case 1:
                        _a = tslib_1.__read.apply(void 0, [_b.sent(), 2]), txn = _a[0], result = _a[1];
                        if (txn.isPending()) {
                            return [2 /*return*/ , result];
                        }
                        return [2 /*return*/ ];
                }
            });
        });
    };
    Blockchain.prototype.createObservedTransaction = function(transaction) {
        try {
            transaction.sendTransaction().then(function(response) {
                var _a = tslib_1.__read(response, 2),
                    txReturned = _a[0],
                    TranID = _a[1];
                txReturned.confirm(TranID).then(function(txConfirmed) {
                    transaction.emitter.resolve(txConfirmed);
                });
            });
            return transaction.emitter;
        } catch (err) {
            throw err;
        }
    };
    Blockchain.prototype.sendRawStakingTransaction = function(staking) {
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var _a, txn, result;
            return tslib_1.__generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        if (!staking.isSigned() || !staking) {
                            throw new Error('staking transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , staking.sendTransaction()];
                    case 1:
                        _a = tslib_1.__read.apply(void 0, [_b.sent(), 2]), txn = _a[0], result = _a[1];
                        if (txn.isPending()) {
                            return [2 /*return*/ , result];
                        }
                        return [2 /*return*/ ];
                }
            });
        });
    };
    Blockchain.prototype.createObservedStakingTransaction = function(staking) {
        try {
            staking.sendTransaction().then(function(response) {
                var _a = tslib_1.__read(response, 2),
                    txReturned = _a[0],
                    TranID = _a[1];
                txReturned.confirm(TranID).then(function(txConfirmed) {
                    staking.emitter.resolve(txConfirmed);
                });
            });
            return staking.emitter;
        } catch (err) {
            throw err;
        }
    };
    Blockchain.prototype.estimateGas = function(_a) {
        var to = _a.to,
            data = _a.data,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_estimateGas" /* EstimateGas */ , [{ to: crypto_1.getAddress(to).checksum, data: data }], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.gasPrice = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_gasPrice" /* GasPrice */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.call = function(_a) {
        var payload = _a.payload,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return tslib_1.__awaiter(this, void 0, void 0, function() {
            var result;
            return tslib_1.__generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("ngy_call" /* Call */ , [payload, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.newPendingTransactions = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof network_1.WSProvider) {
            return new network_1.NewPendingTransactions(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.newBlockHeaders = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof network_1.WSProvider) {
            return new network_1.NewHeaders(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.syncing = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof network_1.WSProvider) {
            return new network_1.Syncing(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.logs = function(options, shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof network_1.WSProvider) {
            return new network_1.LogSub(options, this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    tslib_1.__decorate([
        utils_1.assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBalance", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            returnObject: ['isBoolean', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockByHash", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            returnObject: ['isBoolean', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockByNumber", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockTransactionCountByHash", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockNumber: ['isBlockNumber', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockTransactionCountByNumber", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            index: ['isHex', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByBlockHashAndIndex", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            index: ['isHex', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByBlockNumberAndIndex", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            txnHash: ['isHash', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByHash", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            txnHash: ['isString', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionReceipt", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            txnHash: ['isString', "required" /* required */ ],
            shardID: ['isNumber', "required" /* required */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getCxReceiptByHash", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getCode", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            position: ['isHex', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getStorageAt", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionCount", null);
    tslib_1.__decorate([
        utils_1.assertObject({
            to: ['isValidAddress', "optional" /* optional */ ],
            data: ['isHex', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", Promise)
    ], Blockchain.prototype, "estimateGas", null);
    return Blockchain;
}());
exports.Blockchain = Blockchain;
