// @nordicenergy-js/core/dist/index.esm.js
/**
 * Test Banner
 */

import { ResponseMiddleware, WSProvider, NewPendingTransactions, NewHeaders, LogSub, Syncing, Provider, HttpProvider, Messenger } from '@nordicenergy-js/network';
import * as utils from '@nordicenergy-js/utils';
import { assertObject, Unit, defaultConfig, NordicEnergyCore, hexToNumber } from '@nordicenergy-js/utils';
import * as crypto from '@nordicenergy-js/crypto';
import { getAddress } from '@nordicenergy-js/crypto';
import { TransactionFactory, Transaction } from '@nordicenergy-js/transaction';
import { StakingTransaction, StakingFactory } from '@nordicenergy-js/staking';
import { ContractFactory, Contract } from '@nordicenergy-js/contract';
import { Wallet, Account, HDNode } from '@nordicenergy-js/account';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] }
            instanceof Array && function(d, b) { d.__proto__ = b; }) ||
        function(d, b) {
            for (var p in b)
                if (b.hasOwnProperty(p)) d[p] = b[p];
        };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);

    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __decorate(decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}

function __metadata(metadataKey, metadataValue) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
}

function __awaiter(thisArg, _arguments, P, generator) {
    return new(P || (P = Promise))(function(resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }

        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }

        function step(result) { result.dNet ? resolve(result.value) : new P(function(resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] },
        f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;

    function verb(n) { return function(v) { return step([n, v]); }; }

    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).dNet) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return { value: op[1], dNet: false };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [0];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [6, e];
            y = 0;
        } finally { f = t = 0; }
        if (op[0] & 5) throw op[1];
        return { value: op[0] ? op[1] : void 0, dNet: true };
    }
}

function __values(o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator],
        i = 0;
    if (m) return m.call(o);
    return {
        next: function() {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], dNet: !o };
        }
    };
}

function __read(o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o),
        r, ar = [],
        e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).dNet) ar.push(r.value);
    } catch (error) { e = { error: error }; } finally {
        try {
            if (r && !r.dNet && (m = i["return"])) m.call(i);
        } finally { if (e) throw e.error; }
    }
    return ar;
}

function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++)
        ar = ar.concat(__read(arguments[i]));
    return ar;
}

var Blockchain = /** @class */ (function() {
    function Blockchain(messenger) {
        this.messenger = messenger;
    }
    Blockchain.prototype.setMessenger = function(messenger) {
        this.messenger = messenger;
    };
    Blockchain.prototype.getRpcResult = function(result) {
        if (result instanceof ResponseMiddleware) {
            return result.getRaw;
        } else {
            return result;
        }
    };
    /**
     *
     */
    Blockchain.prototype.getBalance = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getBalance" /* GetBalance */ , [getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockNumber = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_blockNumber" /* BlockNumber */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getBlockByHash = function(_a) {
        var blockHash = _a.blockHash,
            _b = _a.returnObject,
            returnObject = _b === void 0 ? true : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getBlockByHash" /* GetBlockByHash */ , [blockHash, returnObject], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getBlockByNumber = function(_a) {
        var _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.returnObject,
            returnObject = _c === void 0 ? true : _c,
            _d = _a.shardID,
            shardID = _d === void 0 ? this.messenger.currentShard : _d;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_e) {
                switch (_e.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getBlockByNumber" /* GetBlockByNumber */ , [blockNumber, returnObject], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _e.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockTransactionCountByHash = function(_a) {
        var blockHash = _a.blockHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getBlockTransactionCountByHash" /* GetBlockTransactionCountByHash */ , [blockHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getBlockTransactionCountByNumber = function(_a) {
        var blockNumber = _a.blockNumber,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getBlockTransactionCountByNumber" /* GetBlockTransactionCountByNumber */ , [blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getTransactionByBlockHashAndIndex = function(_a) {
        var blockHash = _a.blockHash,
            index = _a.index,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionByBlockHashAndIndex" /* GetTransactionByBlockHashAndIndex */ , [blockHash, index], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionByBlockNumberAndIndex = function(_a) {
        var _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            index = _a.index,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionByBlockNumberAndIndex" /* GetTransactionByBlockNumberAndIndex */ , [blockNumber, index], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionByHash = function(_a) {
        var txnHash = _a.txnHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionByHash" /* GetTransactionByHash */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getTransactionReceipt = function(_a) {
        var txnHash = _a.txnHash,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionReceipt" /* GetTransactionReceipt */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getCxReceiptByHash = function(_a) {
        var txnHash = _a.txnHash,
            shardID = _a.shardID;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getCXReceiptByHash" /* GetCXReceiptByHash */ , [txnHash], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    /**
     *
     */
    Blockchain.prototype.getCode = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getCode" /* GetCode */ , [getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.net_peerCount = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("net_peerCount" /* PeerCount */ , [], 'net', shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.net_version = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("net_version" /* NetVersion */ , [], 'net', shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getProtocolVersion = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_protocolVersion" /* ProtocolVersion */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getStorageAt = function(_a) {
        var address = _a.address,
            position = _a.position,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getStorageAt" /* GetStorageAt */ , [getAddress(address).checksum, position, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getTransactionCount = function(_a) {
        var address = _a.address,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionCount" /* GetTransactionCount */ , [getAddress(address).checksum, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.getShardingStructure = function() {
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_getShardingStructure" /* GetShardingStructure */ , [], this.messenger.chainPrefix)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.sendTransaction = function(transaction) {
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        if (!transaction.isSigned() || !transaction) {
                            throw new Error('transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , this.messenger.send("Ngy_sendRawTransaction" /* SendRawTransaction */ , [transaction.getRawTransaction()], this.messenger.chainPrefix, typeof transaction.txParams.shardID === 'string' ?
                            Number.parseInt(transaction.txParams.shardID, 10) :
                            transaction.txParams.shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.sendRawTransaction = function(transaction) {
        return __awaiter(this, void 0, void 0, function() {
            var _a, txn, result;
            return __generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        if (!transaction.isSigned() || !transaction) {
                            throw new Error('transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , transaction.sendTransaction()];
                    case 1:
                        _a = __read.apply(void 0, [_b.sent(), 2]), txn = _a[0], result = _a[1];
                        if (txn.isPending()) {
                            return [2 /*return*/ , result];
                        }
                        return [2 /*return*/ ];
                }
            });
        });
    };
    Blockchain.prototype.createObservedTransaction = function(transaction) {
        try {
            transaction.sendTransaction().then(function(response) {
                var _a = __read(response, 2),
                    txReturned = _a[0],
                    TranID = _a[1];
                txReturned.confirm(TranID).then(function(txConfirmed) {
                    transaction.emitter.resolve(txConfirmed);
                });
            });
            return transaction.emitter;
        } catch (err) {
            throw err;
        }
    };
    Blockchain.prototype.sendRawStakingTransaction = function(staking) {
        return __awaiter(this, void 0, void 0, function() {
            var _a, txn, result;
            return __generator(this, function(_b) {
                switch (_b.label) {
                    case 0:
                        if (!staking.isSigned() || !staking) {
                            throw new Error('staking transaction is not signed or not exist');
                        }
                        return [4 /*yield*/ , staking.sendTransaction()];
                    case 1:
                        _a = __read.apply(void 0, [_b.sent(), 2]), txn = _a[0], result = _a[1];
                        if (txn.isPending()) {
                            return [2 /*return*/ , result];
                        }
                        return [2 /*return*/ ];
                }
            });
        });
    };
    Blockchain.prototype.createObservedStakingTransaction = function(staking) {
        try {
            staking.sendTransaction().then(function(response) {
                var _a = __read(response, 2),
                    txReturned = _a[0],
                    TranID = _a[1];
                txReturned.confirm(TranID).then(function(txConfirmed) {
                    staking.emitter.resolve(txConfirmed);
                });
            });
            return staking.emitter;
        } catch (err) {
            throw err;
        }
    };
    Blockchain.prototype.estimateGas = function(_a) {
        var to = _a.to,
            data = _a.data,
            _b = _a.shardID,
            shardID = _b === void 0 ? this.messenger.currentShard : _b;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_estimateGas" /* EstimateGas */ , [{ to: getAddress(to).checksum, data: data }], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.gasPrice = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_gasPrice" /* GasPrice */ , [], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.call = function(_a) {
        var payload = _a.payload,
            _b = _a.blockNumber,
            blockNumber = _b === void 0 ? "latest" /* latest */ : _b,
            _c = _a.shardID,
            shardID = _c === void 0 ? this.messenger.currentShard : _c;
        return __awaiter(this, void 0, void 0, function() {
            var result;
            return __generator(this, function(_d) {
                switch (_d.label) {
                    case 0:
                        return [4 /*yield*/ , this.messenger.send("Ngy_call" /* Call */ , [payload, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        result = _d.sent();
                        return [2 /*return*/ , this.getRpcResult(result)];
                }
            });
        });
    };
    Blockchain.prototype.newPendingTransactions = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof WSProvider) {
            return new NewPendingTransactions(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.newBlockHeaders = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof WSProvider) {
            return new NewHeaders(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.syncing = function(shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof WSProvider) {
            return new Syncing(this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    Blockchain.prototype.logs = function(options, shardID) {
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        if (this.messenger.provider instanceof WSProvider) {
            return new LogSub(options, this.messenger, shardID);
        } else {
            throw new Error('HttpProvider does not support this feature');
        }
    };
    __decorate([
        assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBalance", null);
    __decorate([
        assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            returnObject: ['isBoolean', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockByHash", null);
    __decorate([
        assertObject({
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            returnObject: ['isBoolean', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockByNumber", null);
    __decorate([
        assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockTransactionCountByHash", null);
    __decorate([
        assertObject({
            blockNumber: ['isBlockNumber', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getBlockTransactionCountByNumber", null);
    __decorate([
        assertObject({
            blockHash: ['isHash', "required" /* required */ ],
            index: ['isHex', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByBlockHashAndIndex", null);
    __decorate([
        assertObject({
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            index: ['isHex', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByBlockNumberAndIndex", null);
    __decorate([
        assertObject({
            txnHash: ['isHash', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionByHash", null);
    __decorate([
        assertObject({
            txnHash: ['isString', "required" /* required */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionReceipt", null);
    __decorate([
        assertObject({
            txnHash: ['isString', "required" /* required */ ],
            shardID: ['isNumber', "required" /* required */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getCxReceiptByHash", null);
    __decorate([
        assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getCode", null);
    __decorate([
        assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            position: ['isHex', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getStorageAt", null);
    __decorate([
        assertObject({
            address: ['isValidAddress', "required" /* required */ ],
            blockNumber: ['isBlockNumber', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "getTransactionCount", null);
    __decorate([
        assertObject({
            to: ['isValidAddress', "optional" /* optional */ ],
            data: ['isHex', "optional" /* optional */ ],
            shardID: ['isNumber', "optional" /* optional */ ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Blockchain.prototype, "estimateGas", null);
    return Blockchain;
}());

var NordicEnergy = /** @class */ (function(_super) {
    __extends(NordicEnergy, _super);

    function NordicEnergy(url, config) {
        if (config === void 0) {
            config = {
                chainId: defaultConfig.Default.Chain_ID,
                chainType: defaultConfig.Default.Chain_Type,
            };
        }
        var _this = _super.call(this, config.chainType, config.chainId) || this;
        _this.Modules = {
            HttpProvider: HttpProvider,
            WSProvider: WSProvider,
            Messenger: Messenger,
            Blockchain: Blockchain,
            TransactionFactory: TransactionFactory,
            StakingFactory: StakingFactory,
            Wallet: Wallet,
            Transaction: Transaction,
            StakingTransaction: StakingTransaction,
            Account: Account,
            Contract: Contract,
        };
        var providerUrl = config.chainUrl || url || defaultConfig.Default.Chain_URL;
        _this.provider = new Provider(providerUrl).provider;
        _this.messenger = new Messenger(_this.provider, _this.chainType, _this.chainId);
        _this.blockchain = new Blockchain(_this.messenger);
        _this.transactions = new TransactionFactory(_this.messenger);
        _this.stakings = new StakingFactory(_this.messenger);
        _this.wallet = new Wallet(_this.messenger);
        _this.contracts = new ContractFactory(_this.wallet);
        _this.crypto = crypto;
        _this.utils = utils;
        _this.defaultShardID = config.shardID;
        if (_this.defaultShardID !== undefined) {
            _this.setShardID(_this.defaultShardID);
        }
        return _this;
    }
    NordicEnergy.prototype.setProvider = function(provider) {
        this.provider = new Provider(provider).provider;
        this.messenger.setProvider(this.provider);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setChainId = function(chainId) {
        this.chainId = chainId;
        this.messenger.setChainId(this.chainId);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setShardID = function(shardID) {
        this.defaultShardID = shardID;
        this.messenger.setDefaultShardID(this.defaultShardID);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setChainType = function(chainType) {
        this.chainType = chainType;
        this.messenger.setChainType(this.chainType);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.shardingStructures = function(shardingStructures) {
        var e_1, _a;
        try {
            for (var shardingStructures_1 = __values(shardingStructures), shardingStructures_1_1 = shardingStructures_1.next(); !shardingStructures_1_1.dNet; shardingStructures_1_1 = shardingStructures_1.next()) {
                var shard = shardingStructures_1_1.value;
                var shardID = typeof shard.shardID === 'string' ? Number.parseInt(shard.shardID, 10) : shard.shardID;
                this.messenger.shardProviders.set(shardID, {
                    current: shard.current !== undefined ? shard.current : false,
                    shardID: shardID,
                    http: shard.http,
                    ws: shard.ws,
                });
            }
        } catch (e_1_1) { e_1 = { error: e_1_1 }; } finally {
            try {
                if (shardingStructures_1_1 && !shardingStructures_1_1.dNet && (_a = shardingStructures_1.return)) _a.call(shardingStructures_1);
            } finally { if (e_1) throw e_1.error; }
        }
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setMessenger = function(messenger) {
        this.blockchain.setMessenger(messenger);
        this.wallet.setMessenger(messenger);
        this.transactions.setMessenger(messenger);
        this.stakings.setMessenger(messenger);
    };
    return NordicEnergy;
}(NordicEnergyCore));

var TruffleProvider = /** @class */ (function(_super) {
    __extends(TruffleProvider, _super);

    function TruffleProvider(provider, hdOptions, chainOptions, transactionOptions) {
        if (provider === void 0) { provider = 'http://localhost:9500'; }
        if (hdOptions === void 0) {
            hdOptions = {
                menmonic: undefined,
                index: 0,
                addressCount: 1,
            };
        }
        if (chainOptions === void 0) {
            chainOptions = {
                shardID: 0,
                chainType: "Ngy" /* NordicEnergy */ ,
                chainId: 2 /* NgyLocal */ ,
            };
        }
        if (transactionOptions === void 0) {
            transactionOptions = {
                gasLimit: '10000000',
                gasPrice: '20000000000',
            };
        }
        var _this = _super.call(this, provider, hdOptions.menmonic, hdOptions.index, hdOptions.addressCount, chainOptions.shardID, chainOptions.chainType, chainOptions.chainId, transactionOptions.gasLimit, transactionOptions.gasPrice) || this;
        _this.resolveResult = function(response) {
            var final = response.getRaw || response;
            delete final.req;
            delete final.responseType;
            return final;
        };
        _this.resolveCallback = function(err, res, callback) {
            try {
                if (err) {
                    callback(err);
                }
                var response = _this.resolveResult(res);
                callback(null, response);
            } catch (error) {
                throw error;
            }
        };
        return _this;
    }
    TruffleProvider.prototype.send = function() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function() {
            var _a, newArgs, id, params, newMethod, callback, _b, accounts, txObj, rawTxn, result, result, result, result;
            var _this = this;
            return __generator(this, function(_c) {
                switch (_c.label) {
                    case 0:
                        _a = this.resolveArgs.apply(this, __spread(args)), newArgs = _a.newArgs, id = _a.id, params = _a.params, newMethod = _a.newMethod, callback = _a.callback;
                        _b = newMethod;
                        switch (_b) {
                            case 'Ngy_accounts':
                                return [3 /*break*/ , 1];
                            case 'Ngy_sendTransaction':
                                return [3 /*break*/ , 2];
                            case 'Ngy_getTransactionReceipt':
                                return [3 /*break*/ , 5];
                            case 'net_version':
                                return [3 /*break*/ , 7];
                            case 'Ngy_getBlockByNumber':
                                return [3 /*break*/ , 8];
                        }
                        return [3 /*break*/ , 10];
                    case 1:
                        {
                            accounts = this.getAccounts();
                            callback(null, {
                                result: accounts,
                                id: id,
                                jsonrpc: '2.0',
                            });
                            return [2 /*return*/ , {
                                result: accounts,
                                id: id,
                                jsonrpc: '2.0',
                            }];
                            // break;
                        }
                        _c.label = 2;
                    case 2:
                        txObj = params[0];
                        return [4 /*yield*/ , this.signTransaction(txObj)];
                    case 3:
                        rawTxn = _c.sent();
                        return [4 /*yield*/ , this.provider.send({
                            id: id,
                            method: 'Ngy_sendRawTransaction',
                            params: [rawTxn],
                            jsonrpc: '2.0',
                        }, function(err, res) { return _this.resolveCallback(err, res, callback); })];
                    case 4:
                        result = _c.sent();
                        return [2 /*return*/ , this.resolveResult(result)];
                    case 5:
                        return [4 /*yield*/ , this.provider.send({
                            id: id,
                            method: 'Ngy_getTransactionReceipt',
                            params: [params[0]],
                            jsonrpc: '2.0',
                        }, function(err, res) {
                            try {
                                if (err) {
                                    callback(err);
                                }
                                var response = _this.resolveResult(res);
                                if (response.result !== null) {
                                    response.result.status = '0x1';
                                }
                                callback(null, response);
                            } catch (error) {
                                throw error;
                            }
                        })];
                    case 6:
                        result = _c.sent();
                        return [2 /*return*/ , this.resolveResult(result)];
                    case 7:
                        {
                            callback(null, {
                                result: String(this.messenger.chainId),
                                id: id,
                                jsonrpc: '2.0',
                            });
                            return [2 /*return*/ , {
                                result: String(this.messenger.chainId),
                                id: id,
                                jsonrpc: '2.0',
                            }];
                        }
                        _c.label = 8;
                    case 8:
                        return [4 /*yield*/ , this.provider.send(newArgs, function(err, res) {
                            try {
                                if (err) {
                                    callback(err);
                                }
                                var response = _this.resolveResult(res);
                                if (new Unit(response.result.gasLimit)
                                    .asWei()
                                    .toWei()
                                    .gt(new Unit(_this.gasLimit).asWei().toWei())) {
                                    response.result.gasLimit = "0x" + new Unit(_this.gasLimit)
                                        .asWei()
                                        .toWei()
                                        .toString('hex');
                                }
                                callback(null, response);
                            } catch (error) {
                                throw error;
                            }
                        })];
                    case 9:
                        result = _c.sent();
                        return [2 /*return*/ , this.resolveResult(result)];
                    case 10:
                        return [4 /*yield*/ , this.provider.send(newArgs, function(err, res) { return _this.resolveCallback(err, res, callback); })];
                    case 11:
                        result = _c.sent();
                        return [2 /*return*/ , this.resolveResult(result)];
                }
            });
        });
    };
    TruffleProvider.prototype.sendAsync = function() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return this.send.apply(this, __spread(args));
    };
    TruffleProvider.prototype.resolveArgs = function() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var method = args[0].method;
        var params = args[0].params;
        var newMethod = method;
        if (method.startsWith('eth')) {
            newMethod = method.replace('eth', 'Ngy');
        }
        args[0].method = newMethod;
        var id = args[0].id;
        return {
            newArgs: args[0],
            id: id,
            params: params,
            newMethod: newMethod,
            callback: args[1],
        };
    };
    return TruffleProvider;
}(HDNode));

var ExtensionType;
(function(ExtensionType) {
    ExtensionType["MathWallet"] = "MathWallet";
})(ExtensionType || (ExtensionType = {}));
var NordicEnergyExtension = /** @class */ (function() {
    function NordicEnergyExtension(wallet, config) {
        if (config === void 0) {
            config = {
                chainId: defaultConfig.Default.Chain_ID,
                chainType: defaultConfig.Default.Chain_Type,
            };
        }
        this.extensionType = null;
        this.wallet = wallet;
        // check if it is mathwallet
        this.isExtension(this.wallet);
        if (wallet.messenger) {
            this.provider = wallet.messenger.provider;
            this.messenger = wallet.messenger;
        } else {
            this.provider = new Provider(config.chainUrl || wallet.network.chain_url).provider;
            this.messenger = new Messenger(this.provider, config.chainType, config.chainId);
        }
        this.wallet.messenger = this.messenger;
        this.blockchain = new Blockchain(this.messenger);
        this.transactions = new TransactionFactory(this.messenger);
        this.contracts = new ContractFactory(this.wallet);
        this.crypto = crypto;
        this.utils = utils;
    }
    NordicEnergyExtension.prototype.setProvider = function(provider) {
        this.provider = new Provider(provider).provider;
        this.messenger.setProvider(this.provider);
        this.setMessenger(this.messenger);
    };
    NordicEnergyExtension.prototype.setShardID = function(shardID) {
        this.defaultShardID = shardID;
        this.messenger.setDefaultShardID(this.defaultShardID);
        this.setMessenger(this.messenger);
    };
    NordicEnergyExtension.prototype.isExtension = function(wallet) {
        var _this = this;
        var isExtension = false;
        this.extensionType = null;
        if (wallet.isMathWallet) {
            isExtension = true;
            this.extensionType = ExtensionType.MathWallet;
            // remake signTransaction of MathWallet
            var signTransaction_1 = this.wallet.signTransaction;
            this.wallet.signTransaction = function(transaction, updateNonce, encodeMode, blockNumber) {
                if (updateNonce === void 0) { updateNonce = true; }
                if (encodeMode === void 0) { encodeMode = 'rlp'; }
                if (blockNumber === void 0) { blockNumber = 'latest'; }
                return __awaiter(_this, void 0, void 0, function() {
                    var extensionAccount, nonce;
                    return __generator(this, function(_a) {
                        switch (_a.label) {
                            case 0:
                                return [4 /*yield*/ , this.wallet.getAccount()];
                            case 1:
                                extensionAccount = _a.sent();
                                if (!updateNonce) return [3 /*break*/ , 3];
                                return [4 /*yield*/ , this.messenger.send("Ngy_getTransactionCount" /* GetTransactionCount */ , [getAddress(extensionAccount.address).checksum, blockNumber], this.messenger.chainPrefix, typeof transaction.txParams.shardID === 'string' ?
                                    Number.parseInt(transaction.txParams.shardID, 10) :
                                    transaction.txParams.shardID)];
                            case 2:
                                nonce = _a.sent();
                                transaction.setParams(__assign(__assign({}, transaction.txParams), { from: getAddress(extensionAccount.address).bech32, nonce: Number.parseInt(hexToNumber(nonce.result), 10) }));
                                return [3 /*break*/ , 4];
                            case 3:
                                transaction.setParams(__assign(__assign({}, transaction.txParams), { from: getAddress(extensionAccount.address).bech32 }));
                                _a.label = 4;
                            case 4:
                                return [2 /*return*/ , signTransaction_1(transaction, false, encodeMode, blockNumber)];
                        }
                    });
                });
            };
        }
        if (!isExtension) {
            throw new Error('Extension is not found');
        }
        return;
    };
    NordicEnergyExtension.prototype.login = function() {
        return __awaiter(this, void 0, void 0, function() {
            var account;
            return __generator(this, function(_a) {
                switch (_a.label) {
                    case 0:
                        return [4 /*yield*/ , this.wallet.getAccount()];
                    case 1:
                        account = _a.sent();
                        // Use address
                        return [2 /*return*/ , account];
                }
            });
        });
    };
    NordicEnergyExtension.prototype.shardingStructures = function(shardingStructures) {
        var e_1, _a;
        try {
            for (var shardingStructures_1 = __values(shardingStructures), shardingStructures_1_1 = shardingStructures_1.next(); !shardingStructures_1_1.dNet; shardingStructures_1_1 = shardingStructures_1.next()) {
                var shard = shardingStructures_1_1.value;
                var shardID = typeof shard.shardID === 'string' ? Number.parseInt(shard.shardID, 10) : shard.shardID;
                this.messenger.shardProviders.set(shardID, {
                    current: shard.current !== undefined ? shard.current : false,
                    shardID: shardID,
                    http: shard.http,
                    ws: shard.ws,
                });
            }
        } catch (e_1_1) { e_1 = { error: e_1_1 }; } finally {
            try {
                if (shardingStructures_1_1 && !shardingStructures_1_1.dNet && (_a = shardingStructures_1.return)) _a.call(shardingStructures_1);
            } finally { if (e_1) throw e_1.error; }
        }
        this.setMessenger(this.messenger);
    };
    NordicEnergyExtension.prototype.setMessenger = function(messenger) {
        this.blockchain.setMessenger(messenger);
        this.wallet.messenger = messenger;
        this.transactions.setMessenger(messenger);
    };
    return NordicEnergyExtension;
}());

var UrlType;
(function(UrlType) {
    UrlType[UrlType["http"] = 0] = "http";
    UrlType[UrlType["ws"] = 1] = "ws";
})(UrlType || (UrlType = {}));

export { NordicEnergy, Blockchain, TruffleProvider, ExtensionType, NordicEnergyExtension, UrlType };
