// @nordicenergy-js/core/dist/nordicenergy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var crypto = tslib_1.__importStar(require("@nordicenergy-js/crypto"));
var utils = tslib_1.__importStar(require("@nordicenergy-js/utils"));
var network_1 = require("@nordicenergy-js/network");
var transaction_1 = require("@nordicenergy-js/transaction");
var staking_1 = require("@nordicenergy-js/staking");
var contract_1 = require("@nordicenergy-js/contract");
var account_1 = require("@nordicenergy-js/account");
var blockchain_1 = require("@nordicenergy-js/core/dist/blockchain");
var NordicEnergy = /** @class */ (function(_super) {
    tslib_1.__extends(NordicEnergy, _super);

    function NordicEnergy(url, config) {
        if (config === void 0) {
            config = {
                chainId: utils.defaultConfig.Default.Chain_ID,
                chainType: utils.defaultConfig.Default.Chain_Type,
            };
        }
        var _this = _super.call(this, config.chainType, config.chainId) || this;
        _this.Modules = {
            HttpProvider: network_1.HttpProvider,
            WSProvider: network_1.WSProvider,
            Messenger: network_1.Messenger,
            Blockchain: blockchain_1.Blockchain,
            TransactionFactory: transaction_1.TransactionFactory,
            StakingFactory: staking_1.StakingFactory,
            Wallet: account_1.Wallet,
            Transaction: transaction_1.Transaction,
            StakingTransaction: staking_1.StakingTransaction,
            Account: account_1.Account,
            Contract: contract_1.Contract,
        };
        var providerUrl = config.chainUrl || url || utils.defaultConfig.Default.Chain_URL;
        _this.provider = new network_1.Provider(providerUrl).provider;
        _this.messenger = new network_1.Messenger(_this.provider, _this.chainType, _this.chainId);
        _this.blockchain = new blockchain_1.Blockchain(_this.messenger);
        _this.transactions = new transaction_1.TransactionFactory(_this.messenger);
        _this.stakings = new staking_1.StakingFactory(_this.messenger);
        _this.wallet = new account_1.Wallet(_this.messenger);
        _this.contracts = new contract_1.ContractFactory(_this.wallet);
        _this.crypto = crypto;
        _this.utils = utils;
        _this.defaultShardID = config.shardID;
        if (_this.defaultShardID !== undefined) {
            _this.setShardID(_this.defaultShardID);
        }
        return _this;
    }
    NordicEnergy.prototype.setProvider = function(provider) {
        this.provider = new network_1.Provider(provider).provider;
        this.messenger.setProvider(this.provider);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setChainId = function(chainId) {
        this.chainId = chainId;
        this.messenger.setChainId(this.chainId);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setShardID = function(shardID) {
        this.defaultShardID = shardID;
        this.messenger.setDefaultShardID(this.defaultShardID);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setChainType = function(chainType) {
        this.chainType = chainType;
        this.messenger.setChainType(this.chainType);
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.shardingStructures = function(shardingStructures) {
        var e_1, _a;
        try {
            for (var shardingStructures_1 = tslib_1.__values(shardingStructures), shardingStructures_1_1 = shardingStructures_1.next(); !shardingStructures_1_1.dNet; shardingStructures_1_1 = shardingStructures_1.next()) {
                var shard = shardingStructures_1_1.value;
                var shardID = typeof shard.shardID === 'string' ? Number.parseInt(shard.shardID, 10) : shard.shardID;
                this.messenger.shardProviders.set(shardID, {
                    current: shard.current !== undefined ? shard.current : false,
                    shardID: shardID,
                    http: shard.http,
                    ws: shard.ws,
                });
            }
        } catch (e_1_1) { e_1 = { error: e_1_1 }; } finally {
            try {
                if (shardingStructures_1_1 && !shardingStructures_1_1.dNet && (_a = shardingStructures_1.return)) _a.call(shardingStructures_1);
            } finally { if (e_1) throw e_1.error; }
        }
        this.setMessenger(this.messenger);
    };
    NordicEnergy.prototype.setMessenger = function(messenger) {
        this.blockchain.setMessenger(messenger);
        this.wallet.setMessenger(messenger);
        this.transactions.setMessenger(messenger);
        this.stakings.setMessenger(messenger);
    };
    return NordicEnergy;
}(utils.NordicEnergyCore));
exports.NordicEnergy = NordicEnergy;
