// @nordicenergy-js/network/dist/rpcMethod/rpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RPCMethod;
(function (RPCMethod) {
    // 1. Ngy_getBlockByHash
    RPCMethod["GetBlockByHash"] = "Ngy_getBlockByHash";
    // 2. Ngy_getBlockByNumber
    RPCMethod["GetBlockByNumber"] = "Ngy_getBlockByNumber";
    // 3. Ngy_getBlockTransactionCountByHash
    RPCMethod["GetBlockTransactionCountByHash"] = "Ngy_getBlockTransactionCountByHash";
    // 4. Ngy_getBlockTransactionCountByNumber
    RPCMethod["GetBlockTransactionCountByNumber"] = "Ngy_getBlockTransactionCountByNumber";
    // 5. Ngy_getCode
    RPCMethod["GetCode"] = "Ngy_getCode";
    // 6. Ngy_getTransactionByBlockHashAndIndex
    RPCMethod["GetTransactionByBlockHashAndIndex"] = "Ngy_getTransactionByBlockHashAndIndex";
    // 7. Ngy_getTransactionByBlockNumberAndIndex
    RPCMethod["GetTransactionByBlockNumberAndIndex"] = "Ngy_getTransactionByBlockNumberAndIndex";
    // 8. Ngy_getTransactionByHash
    RPCMethod["GetTransactionByHash"] = "Ngy_getTransactionByHash";
    RPCMethod["GetTransactionReceipt"] = "Ngy_getTransactionReceipt";
    RPCMethod["GetCXReceiptByHash"] = "Ngy_getCXReceiptByHash";
    // 9. Ngy_syncing
    RPCMethod["Syncing"] = "Ngy_syncing";
    // 10. net_peerCount
    RPCMethod["PeerCount"] = "net_peerCount";
    // 11. Ngy_getBalance
    RPCMethod["GetBalance"] = "Ngy_getBalance";
    // 12. Ngy_getStorageAt
    RPCMethod["GetStorageAt"] = "Ngy_getStorageAt";
    // 13. Ngy_getTransactionCount
    RPCMethod["GetTransactionCount"] = "Ngy_getTransactionCount";
    // 14. Ngy_sendTransaction
    RPCMethod["SendTransaction"] = "Ngy_sendTransaction";
    // 15. Ngy_sendRawTransaction
    RPCMethod["SendRawTransaction"] = "Ngy_sendRawTransaction";
    // 16. Ngy_subscribe
    RPCMethod["Subscribe"] = "Ngy_subscribe";
    // 17. Ngy_getlogs
    RPCMethod["GetPastLogs"] = "Ngy_getLogs";
    // 18. Ngy_getWork
    RPCMethod["GetWork"] = "Ngy_getWork";
    // 19. Ngy_submitWork
    // SubmitWork = 'Ngy_submitWork',
    // 20. Ngy_getProof
    RPCMethod["GetProof"] = "Ngy_getProof";
    // 21, Ngy_getFilterChanges
    RPCMethod["GetFilterChanges"] = "Ngy_getFilterChanges";
    // 22. Ngy_newPendingTransactionFilter
    RPCMethod["NewPendingTransactionFilter"] = "Ngy_newPendingTransactionFilter";
    // 23. Ngy_newBlockFilter
    RPCMethod["NewBlockFilter"] = "Ngy_newBlockFilter";
    // 24. Ngy_newFilter
    RPCMethod["NewFilter"] = "Ngy_newFilter";
    // 25. Ngy_call
    RPCMethod["Call"] = "Ngy_call";
    // 26. Ngy_estimateGas
    RPCMethod["EstimateGas"] = "Ngy_estimateGas";
    // 27. Ngy_gasPrice
    RPCMethod["GasPrice"] = "Ngy_gasPrice";
    // 28. Ngy_blockNumber
    RPCMethod["BlockNumber"] = "Ngy_blockNumber";
    // 29. Ngy_unsubscribe
    RPCMethod["UnSubscribe"] = "Ngy_unsubscribe";
    // 30. net_version
    RPCMethod["NetVersion"] = "net_version";
    // 31. Ngy_protocolVersion
    RPCMethod["ProtocolVersion"] = "Ngy_protocolVersion";
    // 32. Ngy_getShardingStructure
    RPCMethod["GetShardingStructure"] = "Ngy_getShardingStructure";
    // 33. Ngy_sendRawStakingTransaction
    RPCMethod["SendRawStakingTransaction"] = "Ngy_sendRawStakingTransaction";
})(RPCMethod = exports.RPCMethod || (exports.RPCMethod = {}));
var RPCErrorCode;
(function (RPCErrorCode) {
    // Standard JSON-RPC 2.0 errors
    // RPC_INVALID_REQUEST is internally mapped to HTTP_BAD_REQUEST (400).
    // It should not be used for application-layer errors.
    RPCErrorCode[RPCErrorCode["RPC_INVALID_REQUEST"] = -32600] = "RPC_INVALID_REQUEST";
    // RPC_METHOD_NOT_FOUND is internally mapped to HTTP_NOT_FOUND (404).
    // It should not be used for application-layer errors.
    RPCErrorCode[RPCErrorCode["RPC_METHOD_NOT_FOUND"] = -32601] = "RPC_METHOD_NOT_FOUND";
    RPCErrorCode[RPCErrorCode["RPC_INVALID_PARAMS"] = -32602] = "RPC_INVALID_PARAMS";
    // RPC_INTERNAL_ERROR should only be used for genuine errors in bitcoind
    // (for example datadir corruption).
    RPCErrorCode[RPCErrorCode["RPC_INTERNAL_ERROR"] = -32603] = "RPC_INTERNAL_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_PARSE_ERROR"] = -32700] = "RPC_PARSE_ERROR";
    // General application defined errors
    RPCErrorCode[RPCErrorCode["RPC_MISC_ERROR"] = -1] = "RPC_MISC_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_TYPE_ERROR"] = -3] = "RPC_TYPE_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_INVALID_ADDRESS_OR_KEY"] = -5] = "RPC_INVALID_ADDRESS_OR_KEY";
    RPCErrorCode[RPCErrorCode["RPC_INVALID_PARAMETER"] = -8] = "RPC_INVALID_PARAMETER";
    RPCErrorCode[RPCErrorCode["RPC_DATABASE_ERROR"] = -20] = "RPC_DATABASE_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_DESERIALIZATION_ERROR"] = -22] = "RPC_DESERIALIZATION_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_VERIFY_ERROR"] = -25] = "RPC_VERIFY_ERROR";
    RPCErrorCode[RPCErrorCode["RPC_VERIFY_REJECTED"] = -26] = "RPC_VERIFY_REJECTED";
    RPCErrorCode[RPCErrorCode["RPC_IN_WARMUP"] = -28] = "RPC_IN_WARMUP";
    RPCErrorCode[RPCErrorCode["RPC_METHOD_DEPRECATED"] = -32] = "RPC_METHOD_DEPRECATED";
})(RPCErrorCode = exports.RPCErrorCode || (exports.RPCErrorCode = {}));
