// @harmony-js/account/dist/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var crypto_1 = require("@harmony-js/crypto");
var utils_1 = require("@harmony-js/utils");
var transaction_1 = require("@harmony-js/transaction");
var utils_2 = require("./utils");
var Account = /** @class */ (function () {
    function Account(key, messenger) {
        if (messenger === void 0) { messenger = utils_2.defaultMessenger; }
        this.balance = '0';
        this.nonce = 0;
        this.encrypted = false;
        this.messenger = messenger;
        if (!key) {
            this._new();
        }
        else {
            this._import(key);
        }
        this.shardID = this.messenger.currentShard || 0;
        this.shards = new Map();
        this.shards.set(this.shardID, {
            address: "" + this.bech32Address + utils_1.AddressSuffix + "0",
            balance: this.balance || '0',
            nonce: this.nonce || 0,
        });
    }
    /**
     * @function new static method create account
     * @return {Account} Account instance
     */
    Account.new = function () {
        var newAcc = new Account()._new();
        return newAcc;
    };
    /**
     * @function add static method add a private key to Account
     * @param  {string} key - private Key
     * @return {Account} Account instance
     */
    Account.add = function (key) {
        var newAcc = new Account()._import(key);
        return newAcc;
    };
    Object.defineProperty(Account.prototype, "checksumAddress", {
        /**
         * @function checksumAddress checsumAddress getter
         * @return {string} get the checksumAddress
         */
        get: function () {
            return this.address ? crypto_1.getAddress(this.address).checksum : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Account.prototype, "bech32Address", {
        get: function () {
            return this.address ? crypto_1.getAddress(this.address).bech32 : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Account.prototype, "bech32TestNetAddress", {
        get: function () {
            return this.address ? crypto_1.getAddress(this.address).bech32TestNet : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Account.prototype, "getShardsCount", {
        /**
         * @function getShardsCount getShards number with this Account
         * @return {number} shard size
         */
        get: function () {
            return this.shards.size;
        },
        enumerable: true,
        configurable: true
    });
    Account.prototype.toFile = function (password, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var file;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.privateKey && utils_1.isPrivateKey(this.privateKey))) return [3 /*break*/, 2];
                        return [4 /*yield*/, crypto_1.encrypt(this.privateKey, password, options)];
                    case 1:
                        file = _a.sent();
                        this.privateKey = file;
                        this.encrypted = true;
                        return [2 /*return*/, file];
                    case 2: throw new Error('Encryption failed because PrivateKey is not correct');
                }
            });
        });
    };
    Account.prototype.fromFile = function (keyStore, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var file, decyptedPrivateKey, error_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        if (!password) {
                            throw new Error('you must provide password');
                        }
                        file = JSON.parse(keyStore);
                        return [4 /*yield*/, crypto_1.decrypt(file, password)];
                    case 1:
                        decyptedPrivateKey = _a.sent();
                        if (utils_1.isPrivateKey(decyptedPrivateKey)) {
                            return [2 /*return*/, this._import(decyptedPrivateKey)];
                        }
                        else {
                            throw new Error('decrypted failed');
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        throw error_1;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @function getBalance get Account's balance
     * @return {type} {description}
     */
    Account.prototype.getBalance = function (blockNumber) {
        if (blockNumber === void 0) { blockNumber = 'latest'; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var balance, nonce, error_2;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        if (!this.messenger) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.messenger.send("hmy_getBalance" /* GetBalance */, [this.address, blockNumber], this.messenger.chainPrefix, this.messenger.currentShard || 0)];
                    case 1:
                        balance = _a.sent();
                        return [4 /*yield*/, this.messenger.send("hmy_getTransactionCount" /* GetTransactionCount */, [this.address, blockNumber], this.messenger.chainPrefix, this.messenger.currentShard || 0)];
                    case 2:
                        nonce = _a.sent();
                        if (balance.isError()) {
                            throw balance.error.message;
                        }
                        if (nonce.isError()) {
                            throw nonce.error.message;
                        }
                        this.balance = utils_1.hexToNumber(balance.result);
                        this.nonce = Number.parseInt(utils_1.hexToNumber(nonce.result), 10);
                        this.shardID = this.messenger.currentShard || 0;
                        return [3 /*break*/, 4];
                    case 3: throw new Error('No Messenger found');
                    case 4: return [2 /*return*/, {
                            balance: this.balance,
                            nonce: this.nonce,
                            shardID: this.shardID,
                        }];
                    case 5:
                        error_2 = _a.sent();
                        throw error_2;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @function updateShards
     * @return {Promise<string>} {description}
     */
    Account.prototype.updateBalances = function (blockNumber) {
        if (blockNumber === void 0) { blockNumber = 'latest'; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var shardProviders, shardProviders_1, shardProviders_1_1, _a, name_1, val, balanceObject, e_1_1, currentShard;
            var e_1, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        shardProviders = this.messenger.shardProviders;
                        if (!(shardProviders.size > 1)) return [3 /*break*/, 10];
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 7, 8, 9]);
                        shardProviders_1 = tslib_1.__values(shardProviders), shardProviders_1_1 = shardProviders_1.next();
                        _c.label = 2;
                    case 2:
                        if (!!shardProviders_1_1.done) return [3 /*break*/, 6];
                        _a = tslib_1.__read(shardProviders_1_1.value, 2), name_1 = _a[0], val = _a[1];
                        return [4 /*yield*/, this.getShardBalance(val.shardID, blockNumber)];
                    case 3:
                        balanceObject = _c.sent();
                        return [4 /*yield*/, this.shards.set(name_1 === val.shardID ? name_1 : val.shardID, balanceObject)];
                    case 4:
                        _c.sent();
                        _c.label = 5;
                    case 5:
                        shardProviders_1_1 = shardProviders_1.next();
                        return [3 /*break*/, 2];
                    case 6: return [3 /*break*/, 9];
                    case 7:
                        e_1_1 = _c.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 9];
                    case 8:
                        try {
                            if (shardProviders_1_1 && !shardProviders_1_1.done && (_b = shardProviders_1.return)) _b.call(shardProviders_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 9: return [3 /*break*/, 12];
                    case 10: return [4 /*yield*/, this.getShardBalance(this.messenger.currentShard || 0, blockNumber)];
                    case 11:
                        currentShard = _c.sent();
                        this.shards.set(this.messenger.currentShard || 0, currentShard);
                        _c.label = 12;
                    case 12: return [2 /*return*/];
                }
            });
        });
    };
    Account.prototype.signTransaction = function (transaction, updateNonce, encodeMode, blockNumber) {
        if (updateNonce === void 0) { updateNonce = true; }
        if (encodeMode === void 0) { encodeMode = 'rlp'; }
        if (blockNumber === void 0) { blockNumber = 'latest'; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var txShardID, shardBalanceObject, shardNonce, _a, signature_1, rawTransaction_1;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.privateKey || !utils_1.isPrivateKey(this.privateKey)) {
                            throw new Error(this.privateKey + " is not found or not correct");
                        }
                        if (!updateNonce) return [3 /*break*/, 2];
                        txShardID = transaction.txParams.shardID;
                        return [4 /*yield*/, this.getShardBalance(typeof txShardID === 'string' ? Number.parseInt(txShardID, 10) : txShardID, blockNumber)];
                    case 1:
                        shardBalanceObject = _b.sent();
                        if (shardBalanceObject !== undefined) {
                            shardNonce = shardBalanceObject.nonce;
                            transaction.setParams(tslib_1.__assign(tslib_1.__assign({}, transaction.txParams), { from: this.messenger.chainPrefix === "hmy" /* Harmony */
                                    ? this.bech32Address
                                    : this.checksumAddress || '0x', nonce: shardNonce }));
                        }
                        else {
                            transaction.setParams(tslib_1.__assign(tslib_1.__assign({}, transaction.txParams), { from: this.messenger.chainPrefix === "hmy" /* Harmony */
                                    ? this.bech32Address
                                    : this.checksumAddress || '0x', nonce: 0 }));
                        }
                        _b.label = 2;
                    case 2:
                        if (encodeMode === 'rlp') {
                            _a = tslib_1.__read(transaction_1.RLPSign(transaction, this.privateKey), 2), signature_1 = _a[0], rawTransaction_1 = _a[1];
                            return [2 /*return*/, transaction.map(function (obj) {
                                    return tslib_1.__assign(tslib_1.__assign({}, obj), { signature: signature_1,
                                        rawTransaction: rawTransaction_1, from: _this.messenger.chainPrefix === "hmy" /* Harmony */
                                            ? _this.bech32Address
                                            : _this.checksumAddress || '0x' });
                                })];
                        }
                        else {
                            // TODO: if we use other encode method, eg. protobuf, we should implement this
                            return [2 /*return*/, transaction];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Account.prototype.signStaking = function (staking, updateNonce, encodeMode, blockNumber, shardID) {
        if (updateNonce === void 0) { updateNonce = true; }
        if (encodeMode === void 0) { encodeMode = 'rlp'; }
        if (blockNumber === void 0) { blockNumber = 'latest'; }
        if (shardID === void 0) { shardID = this.messenger.currentShard; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var txShardID, shardBalanceObject, shardNonce, _a, signature, rawTransaction;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.privateKey || !utils_1.isPrivateKey(this.privateKey)) {
                            throw new Error(this.privateKey + " is not found or not correct");
                        }
                        if (!updateNonce) return [3 /*break*/, 2];
                        txShardID = shardID;
                        return [4 /*yield*/, this.getShardBalance(typeof txShardID === 'string' ? Number.parseInt(txShardID, 10) : txShardID, blockNumber)];
                    case 1:
                        shardBalanceObject = _b.sent();
                        if (shardBalanceObject !== undefined) {
                            shardNonce = shardBalanceObject.nonce;
                            staking.setFromAddress(this.messenger.chainPrefix === "hmy" /* Harmony */
                                ? this.bech32Address
                                : this.checksumAddress || '0x');
                            staking.setNonce(shardNonce);
                        }
                        else {
                            staking.setFromAddress(this.messenger.chainPrefix === "hmy" /* Harmony */
                                ? this.bech32Address
                                : this.checksumAddress || '0x');
                            staking.setNonce(0);
                        }
                        _b.label = 2;
                    case 2:
                        if (encodeMode === 'rlp') {
                            _a = tslib_1.__read(staking.rlpSign(this.privateKey), 2), signature = _a[0], rawTransaction = _a[1];
                            staking.setRawTransaction(rawTransaction);
                            staking.setSignature(signature);
                            staking.setFromAddress(this.messenger.chainPrefix === "hmy" /* Harmony */
                                ? this.bech32Address
                                : this.checksumAddress || '0x');
                            return [2 /*return*/, staking];
                        }
                        else {
                            // TODO: if we use other encode method, eg. protobuf, we should implement this
                            return [2 /*return*/, staking];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Account.prototype.setMessenger = function (messenger) {
        this.messenger = messenger;
    };
    Account.prototype.getAddressFromShardID = function (shardID) {
        var shardObject = this.shards.get(shardID);
        if (shardObject) {
            return shardObject.address;
        }
        else {
            return undefined;
        }
    };
    Account.prototype.getAddresses = function () {
        var e_2, _a;
        var addressArray = [];
        try {
            for (var _b = tslib_1.__values(this.shards), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = tslib_1.__read(_c.value, 2), name_2 = _d[0], val = _d[1];
                var index = typeof name_2 === 'string' ? Number.parseInt(name_2, 10) : name_2;
                addressArray[index] = val.address;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return addressArray;
    };
    Account.prototype.getShardBalance = function (shardID, blockNumber) {
        if (blockNumber === void 0) { blockNumber = 'latest'; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var balance, nonce;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.messenger.send("hmy_getBalance" /* GetBalance */, [this.address, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 1:
                        balance = _a.sent();
                        return [4 /*yield*/, this.messenger.send("hmy_getTransactionCount" /* GetTransactionCount */, [this.address, blockNumber], this.messenger.chainPrefix, shardID)];
                    case 2:
                        nonce = _a.sent();
                        if (balance.isError()) {
                            throw balance.error.message;
                        }
                        if (nonce.isError()) {
                            throw nonce.error.message;
                        }
                        return [2 /*return*/, {
                                address: "" + this.bech32Address + utils_1.AddressSuffix + shardID,
                                balance: utils_1.hexToNumber(balance.result),
                                nonce: Number.parseInt(utils_1.hexToNumber(nonce.result), 10),
                            }];
                }
            });
        });
    };
    /**
     * @function _new private method create Account
     * @return {Account} Account instance
     */
    Account.prototype._new = function () {
        var prv = crypto_1.generatePrivateKey();
        if (!utils_1.isPrivateKey(prv)) {
            throw new Error('key gen failed');
        }
        return this._import(prv);
    };
    /**
     * @function _import private method import a private Key
     * @param  {string} key - private key
     * @return {Account} Account instance
     */
    Account.prototype._import = function (key) {
        if (!utils_1.isPrivateKey(key)) {
            throw new Error(key + " is not PrivateKey");
        }
        this.privateKey = utils_1.add0xToString(key);
        this.publicKey = crypto_1.getPubkeyFromPrivateKey(this.privateKey);
        this.address = crypto_1.getAddressFromPrivateKey(this.privateKey);
        this.shardID = this.messenger.currentShard || 0;
        this.shards = new Map();
        this.shards.set(this.shardID, {
            address: "" + this.bech32Address + utils_1.AddressSuffix + "0",
            balance: this.balance || '0',
            nonce: this.nonce || 0,
        });
        this.encrypted = false;
        return this;
    };
    return Account;
}());
exports.Account = Account;
