// @nordicenergy-js/utils/dist/chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChainType;
(function(ChainType) {
    ChainType["NordicEnergy"] = "ngy";
    ChainType["Ethereum"] = "eth";
})(ChainType = exports.ChainType || (exports.ChainType = {}));
var ChainID;
(function(ChainID) {
    ChainID[ChainID["Default"] = 0] = "Default";
    ChainID[ChainID["EthMainnet"] = 1] = "EthMainnet";
    ChainID[ChainID["Morden"] = 2] = "Morden";
    ChainID[ChainID["Ropsten"] = 3] = "Ropsten";
    ChainID[ChainID["Rinkeby"] = 4] = "Rinkeby";
    ChainID[ChainID["RootstockMainnet"] = 30] = "RootstockMainnet";
    ChainID[ChainID["RootstockTestnet"] = 31] = "RootstockTestnet";
    ChainID[ChainID["Kovan"] = 42] = "Kovan";
    ChainID[ChainID["EtcMainnet"] = 61] = "EtcMainnet";
    ChainID[ChainID["EtcTestnet"] = 62] = "EtcTestnet";
    ChainID[ChainID["Geth"] = 1337] = "Geth";
    ChainID[ChainID["Ganache"] = 0] = "Ganache";
    ChainID[ChainID["NgyMainnet"] = 1] = "NgyMainnet";
    ChainID[ChainID["NgyTestnet"] = 2] = "NgyTestnet";
    ChainID[ChainID["NgyLocal"] = 2] = "NgyLocal";
    ChainID[ChainID["NgyPangaea"] = 3] = "NgyPangaea";
})(ChainID = exports.ChainID || (exports.ChainID = {}));
exports.defaultConfig = {
    Default: {
        Chain_ID: 2 /* NgyLocal */ ,
        Chain_Type: "ngy" /* Nordic Energy */ ,
        Chain_URL: 'http://localhost:9500',
        Network_ID: 'Local',
    },
    DefaultWS: {
        Chain_ID: 2 /* NgyLocal */ ,
        Chain_Type: "ngy" /* Nordic Energy */ ,
        Chain_URL: 'ws://localhost:9800',
        Network_ID: 'LocalWS',
    },
};
var NordicEnergyCore = /** @class */ (function() {
    function NordicEnergyCore(chainType, chainId) {
        if (chainId === void 0) { chainId = exports.defaultConfig.Default.Chain_ID; }
        this.chainType = chainType;
        this.chainId = chainId;
    }
    Object.defineProperty(NordicEnergyCore.prototype, "chainPrefix", {
        get: function() {
            switch (this.chainType) {
                case "eth" /* Ethereum */ :
                    {
                        return 'eth';
                    }
                case "ngy" /* Nordic Energy */ :
                    {
                        return 'ngy';
                    }
                default:
                    {
                        return 'ngy';
                    }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NordicEnergyCore.prototype, "getChainId", {
        get: function() {
            return this.chainId;
        },
        enumerable: true,
        configurable: true
    });
    NordicEnergyCore.prototype.setChainId = function(chainId) {
        this.chainId = chainId;
    };
    NordicEnergyCore.prototype.setChainType = function(chainType) {
        this.chainType = chainType;
    };
    return NordicEnergyCore;
}());
exports.NordicEnergyCore = NordicEnergyCore;
exports.HDPath = "m/44'/1023'/0'/0/";
exports.AddressSuffix = '-';
